### PR TITLE
Electrified grille sparks effect

### PR DESCRIPTION
**Pull request by Slava0135**
## About the PR

Makes electrified grilles spawn sparks

**Media**

Peek 2023-04-06 23-34

**Changelog**

:cl:
- add: Powered electrified grilles now create sparks when touched.

### CODE REVIEW

**Comment by brainfood1183** [comment]
make sure it doesnt do this animation when under a window if you havent already :)

**Comment by Slava0135** [comment]
> make sure it doesnt do this animation when under a window if you havent already :)

It only spawns effect when it can actually zap you (if not wearing gloves) so it's in sync with actual grille state.

**Comment by Slava0135** [comment]
New animation by Aleksh

Peek 2023-04-07 20-43

**Comment by Slava0135** [comment]
made effect to be just a layer + made some optimisations
Peek 2023-04-07 22-16

**Comment by Emisse** [comment]
is this performant if it checks every frame ?

**Comment by metalgearsloth** [comment]
> > is this performant if it checks every frame ?
> 
> not really, but there is no way to know if nodes in node container changed state, and thats bad.

Subscribe to power events?

**Comment by AJCM-git** [comment]
> > > is this performant if it checks every frame ?
> > 
> > 
> > not really, but there is no way to know if nodes in node container changed state, and thats bad.
> 
> Subscribe to power events?

Its a lil tricky because the current electrified implementation is not connected directly to power consumers,  it just picks any wire under the grille and doesnt care if the grille uses power or not  (it supports ApcPowerReceivers but not PowerConsumers), but it can be doable.

**Comment by Slava0135** [comment]
Ok, I will try to make this work better

**Comment by Slava0135** [comment]
So we decided it would be too hard to make it performant with current state of power system.
Instead animation will play when someone touches grille.
Peek 2023-04-10 20-55

**Comment by metalgearsloth** [comment]
In future please don't change stuff that's refactored to file-scoped namespaces as it makes the file-diffs unreadable.

File: Content.Server/Electrocution/ElectrocutionSystem.cs [hunk 1]
```diff
@@ -23,107 +23,137 @@
...
+    {
+        UpdateElectrocutions(frameTime);
+        UpdateState(frameTime);
+    }
 
-        public override void Initialize()
+    private void UpdateElectrocutions(float frameTime)
+    {
+        var query = AllEntityQuery<ElectrocutionComponent, PowerConsumerComponent>();
+        while (query.MoveNext(out var uid, out var electrocution, out var consumer))
```
Comment (metalgearsloth) [inline Content.Server/Electrocution/ElectrocutionSystem.cs:89]:
```suggestion
        var query = EntityQueryEnumerator<ElectrocutionComponent, PowerConsumerComponent>();
        while (query.MoveNext(out var uid, out var electrocution, out var consumer))
```

This shouldn't be enumerating pause entities.

File: Content.Server/Electrocution/ElectrocutionSystem.cs [hunk 2]
```diff
@@ -23,107 +23,137 @@
...
-                    if (actual != null)
-                    {
-                        _adminLogger.Add(LogType.Electrocution,
-                            $"{ToPrettyString(finished.Electrocuting):entity} received {actual.Total:damage} powered electrocution damage from {ToPrettyString(finished.Source):source}");
-                    }
+                    _adminLogger.Add(LogType.Electrocution,
+                        $"{ToPrettyString(electrocution.Electrocuting):entity} received {actual.Total:damage} powered electrocution damage from {ToPrettyString(electrocution.Source):source}");
                 }
+            }
+            EntityManager.DeleteEntity(uid);
```
Comment (metalgearsloth) [inline Content.Server/Electrocution/ElectrocutionSystem.cs:112]:
```suggestion
            }
            QueueDel(uid);
```

Iterating entities while deleting them is a sussy proposition.

File: Content.Server/Electrocution/ElectrocutionSystem.cs [hunk 3]
```diff
@@ -23,107 +23,137 @@
...
+    private void UpdateState(float frameTime)
+    {
+        var query = EntityQueryEnumerator<ActivatedElectrifiedComponent, ElectrifiedComponent, TransformComponent>();
+        while (query.MoveNext(out var uid, out var activated, out var electrified, out var transform))
+        {
+            activated.TimeLeft -= frameTime;
+            if (activated.TimeLeft <= 0 || !IsPowered(uid, electrified, transform)) {
+                _appearance.SetData(uid, ElectrifiedVisuals.IsPowered, false);
+                RemComp<ActivatedElectrifiedComponent>(uid);
             }
```
Comment (metalgearsloth) [inline Content.Server/Electrocution/ElectrocutionSystem.cs:125]:
```suggestion
            if (activated.TimeLeft <= 0 || !IsPowered(uid, electrified, transform))
            {
                _appearance.SetData(uid, ElectrifiedVisuals.IsPowered, false);
                RemCompDeferred<ActivatedElectrifiedComponent>(uid);
            }
```

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/Electrocution/ElectrocutionSystem.cs (hunk 2 above)
Comment (Slava0135) [reply Content.Server/Electrocution/ElectrocutionSystem.cs:112]: Yeah, before that entities were added to the list first and then were removed, maybe it was actually reasonable.

**Review by keronshb — APPROVED** [review]
Seems good to me

**Pull request merged**